src/IO/ReadWriteBufferFromHTTP.h

```diff
@@ -209,29 +209,74 @@ namespace detail
     };
 }
 
-class SessionFactory
+
+/// A short-lived HTTP session pool for one endpoint.
+/// Keeps an unlimited number of sessions for one URI.
+/// If URI changes (redirect), clears the pool.
+/// If method is not GET or HEAD, we avoid pooling altogether, just in case.
+/// The pool must outlive all session pointers created by it.
+///
+/// Session is only reused if it has HTTPSessionReuseTag attached, see comment in HTTPCommon.h
+class LocallyPooledSessionFactory
```

Comment:
could we replace it with `HTTPSessionPool` used for s3 read buffer? it has a mode when new sessions will be allocated without waiting after limit is exceeded, but these excessive sessions will be removed when request is finished. imo it is better behaviour because there is no much sense is keeping bursted session forever - they are likely to expire and do reconnect on the next use anyway.

Reply:
I was too afraid to use it here:
 * It never evicts endpoints from `endpoints_pool`. If the user touches a million different hosts (even invalid ones) without restarting the server, it'll get quite big.
 * For each endpoint, it only evicts sessions when new requests come in. If the user touches 100 K different hosts with 30 sessions each, we'll be holding 300K fds indefinitely, if I'm reading the code right.
 * I don't know enough about HTTP to be confident that matching tuple `<host, port, https, proxy_host, proxy_port, proxy_https, max_connections_per_endpoint, resolve_host, wait_on_pool_size_limit>` is sufficient for reuse in all cases. Maybe URL can have some other connection settings not listed here? Maybe I'm being too paranoid.

These are all fine for S3, where the URLs are mostly under our control and not very diverse, but for arbitrary URLs from the user I chickened out. Guess I'll change it to use the global pool and add a setting to disable pooling altogether.

Reply:
> If the user touches a million different hosts
you're talking about `url()` function? indeed, with s3 I don't think it is possible.